### PR TITLE
Fix overlapping bars on continuous axes

### PR DESCRIPTION
**Pull request by krkshitij**
For more information, see [RFC: Fix overlapping bars on continuous axes](https://github.com/microsoft/fluentui-charting-contrib/pull/87)

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 1]
```diff
@@ -40,6 +40,7 @@ import {
   getBarWidth,
```
Comment (fabricteam) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:40]: ## [🕵 fluentuiv8 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31035&commitId=ca64d029e8c15196035a45858df9a4beb10a5254&env=prod&clientType=FLUENTUI)


  <details><summary><b>react-charting-VerticalBarChart</b> 12 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|react-charting-VerticalBarChart.Date Axis- VBC.default.chromium.png|29820| Changed|
|react-charting-VerticalBarChart.Date Axis- VBC - Dark Mode.default.chromium.png|22622| Changed|
|react-charting-VerticalBarChart.Date Axis- VBC - RTL.default.chromium.png|29820| Changed|
|react-charting-VerticalBarChart.Basic - Secondary Y Axis - RTL.default.chromium.png|13142| Changed|
|react-charting-VerticalBarChart.Stacked - Dark Mode.default.chromium.png|756| Changed|
|react-charting-VerticalBarChart.Basic - Secondary Y Axis.default.chromium.png|13185| Changed|
|react-charting-VerticalBarChart.Basic - Secondary Y Axis - Dark Mode.default.chromium.png|5625| Changed|
|react-charting-VerticalBarChart.Stacked - Date Axis - RTL.default.chromium.png|420| Changed|
|react-charting-VerticalBarChart.Stacked - Date Axis - Dark Mode.default.chromium.png|438| Changed|
|react-charting-VerticalBarChart.Stacked.default.chromium.png|1925| Changed|
|react-charting-VerticalBarChart.Stacked - Date Axis.default.chromium.png|420| Changed|
|react-charting-VerticalBarChart.Stacked - RTL.default.chromium.png|1912| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv8 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 2]
```diff
@@ -532,25 +532,18 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
...
-            ? [
-                containerWidth - this.margins.right! - this._barWidth / 2 - this._domainMargin,
-                this.margins.left! + this._domainMargin + this._barWidth / 2,
-              ]
-            : [
-                this.margins.left! + this._domainMargin + this._barWidth / 2,
-                containerWidth - this.margins.right! - this._barWidth / 2 - this._domainMargin,
-              ],
-        );
+        .domain(this._isRtl ? [lDate, sDate] : [sDate, lDate])
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:542]: for RTL domain is the same but range is reversed.

File: packages/react-charting/src/utilities/utilities.ts [hunk 3]
```diff
@@ -1518,3 +1517,25 @@ export const getScalePadding = (prop: number | undefined, shorthandProp?: number
 export const isScalePaddingDefined = (prop: number | undefined, shorthandProp?: number): boolean => {
   return typeof prop === 'number' || typeof shorthandProp === 'number';
 };
+
+export const getClosestPairDiffAndRange = (data: number[] | Date[]): [number, number] | undefined => {
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1521]: add a bunch of UTs for this function

File: packages/react-charting/src/utilities/utilities.ts [hunk 4]
```diff
@@ -961,9 +961,8 @@ export function domainRangeOfVSBCNumeric(
 ): IDomainNRange {
   const xMin = d3Min(points, (point: IDataPoint) => point.x as number)!;
   const xMax = d3Max(points, (point: IDataPoint) => point.x as number)!;
-  // barWidth / 2 - for to get tick middle of the bar
-  const rMax = margins.left! + barWidth / 2;
-  const rMin = width - margins.right! - barWidth / 2;
+  const rMax = margins.left!;
+  const rMin = width - margins.right!;
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:965]: why this change?

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 5]
```diff
@@ -1088,14 +1056,22 @@ export class VerticalStackedBarChartBase extends React.Component<
           containerWidth - (this.margins.left! + MIN_DOMAIN_MARGIN) - (this.margins.right! + MIN_DOMAIN_MARGIN);
         /** Rate at which the space between the bars changes wrt the bar width */
         const barGapRate = this._xAxisInnerPadding / (1 - this._xAxisInnerPadding);
+        this._barWidth = getBarWidth(this.props.barWidth, this.props.maxBarWidth);
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:1059]: is this function called again after containerwidth is updated inside cartesian chart.
Add a comment about the behavior in the code.

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 2 above)
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:543]: Either way is fine. I modified this to make the changes look similar to the numeric scale.

File: packages/react-charting/src/utilities/utilities.ts (hunk 4 above)
Comment (krkshitij) [reply packages/react-charting/src/utilities/utilities.ts:965]: The `getDomainMargins` function in VerticalBarChart already includes these values in the `margins`. The updated margins are then passed to the axis utilities from CartesianChart, removing the need for a direct dependency on the `barwidth` prop.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 5 above)
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:1060]: Yes, the `getDomainMargins` function is called every time CartesianChart rerenders. But there's a possibility that VerticalBarChart might not have rerendered with CartesianChart. In such cases, the function could use previous `barWidth` values, leading to incorrect results.

**Comment by AtishayMsft** [comment]
The max bar width should be set to 16px.

**Comment by AtishayMsft** [comment]
Can you share the behavior for date axis and numeric axis for these versions.
1. Before 2:1 auto spacing fix.
2. After 2:1 auto spacing fix and before this change.
3. After this change.

File: packages/react-charting/src/utilities/utilities.ts (hunk 3 above)
Comment (krkshitij) [reply packages/react-charting/src/utilities/utilities.ts:1521]: done

**Comment by krkshitij** [comment]
> Can you share the behavior for date axis and numeric axis for these versions for following scenarios - 3 bars, 20 bars, 50 bars

1. Before 2:1 auto spacing fix: https://codepen.io/krkshitij/pen/wvZEZyZ
2. After 2:1 auto spacing fix and before this change: https://codepen.io/krkshitij/pen/JjVaVLy
3. After this change: https://fluentuipr.z22.web.core.windows.net/pull/31035/react-charting/demo/index.html#/examples/verticalbarchart:~:text=33-,VerticalBarChart%20dynamic,-%EE%A2%A7

**Comment by AtishayMsft** [comment]
> The max bar width should be set to 16px.

I meant default bar width here

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 6]
```diff
@@ -771,7 +750,7 @@ export class VerticalStackedBarChartBase extends React.Component<
     );
     const shouldFocusWholeStack = this._toFocusWholeStack(_isHavingLines);
 
-    if (this._xAxisType === XAxisTypes.StringAxis && this.props.barWidth === 'auto') {
+    if (this._xAxisType === XAxisTypes.StringAxis) {
       // Setting the bar width here is safe because there are no dependencies earlier in the code
       // that rely on the width of bars in vertical bar charts with string x-axis.
       this._barWidth = getBarWidth(this.props.barWidth, this.props.maxBarWidth, xBarScale.bandwidth());
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:756]: no change needed in this call to barwidth

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 2 above)
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: reverted the change

File: packages/react-charting/src/utilities/UtilityUnitTests.test.ts [hunk 7]
```diff
@@ -1332,3 +1332,23 @@ test('formatValueWithSIPrefix should format a numeric value with appropriate SI
...
+  it('should return undefined if data length is less than 2', () => {
+    const data: number[] = [1];
+    const result = utils.getClosestPairDiffAndRange(data);
+    expect(result).toBeUndefined();
+  });
+
+  it('should return the minimum difference and range for number data', () => {
+    const data: number[] = [1, 5, 3, 9, 2];
+    const result = utils.getClosestPairDiffAndRange(data);
+    expect(result).toEqual([1, 8]);
```
[the remark this answers is not in the document]
Comment (krkshitij) [reply packages/react-charting/src/utilities/UtilityUnitTests.test.ts:1346]: Numbers in `data` represent x-axis values, and they are always rendered in sorted order. So, the closest pairs are (1, 2) and (2, 3).

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 8]
```diff
@@ -397,15 +384,7 @@ export class VerticalStackedBarChartBase extends React.Component<
...
-            cx={
-              isNumeric
-                ? xScale(circlePoint.xItem.xAxisPoint as number)
-                : isDate
-                ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                  (xBarScale as any)(circlePoint.xItem.xAxisPoint as Date)
-                : // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                  (xBarScale as any)(circlePoint.xItem.xAxisPoint as string)
-            }
+            cx={xScale(circlePoint.xItem.xAxisPoint)}
```
[the remark this answers is not in the document]
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:387]: Instead of typecasting `xScale` as `any` each time, I have defined the argument as `any`

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 6 above)
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:756]: For the string x-axis scenario, the existing bar width calculation works fine. The only change is in the `getBarWidth` utility function.

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 9]
```diff
@@ -973,4 +977,20 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
       (d3Max(this._points, (point: IVerticalBarChartDataPoint) => point.y)! <= 0 && !this._isHavingLine)
     );
   }
+
+  private _calculateAppropriateBarWidth = (containerWidth: number) => {
+    const data = (this.props.data?.map(point => point.x) as number[] | Date[] | undefined) || [];
+    const result = getClosestPairDiffAndRange(data);
```
[the remark this answers is not in the document]
Comment (krkshitij) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: done

**Review by AtishayMsft — APPROVED** [review]
